src/core/abr/select_optimal_representation.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Representation } from "../../manifest";
+import arrayFindIndex from "../../utils/array_find_index";
+
+/**
+ * From the given array of Representations (sorted by bitrate order ascending),
+ * returns the one corresponding to the given optimal, minimum and maximum
+ * bitrates.
+ * @param {Array.<Representation>} representations - The representations array,
+ * sorted in bitrate ascending order.
+ * @param {Number} optimalBitrate - The optimal bitrate the Representation
+ * should have under the current condition.
+ * @param {Number} minBitrate - The minimum bitrate the choosen Representation
+ * should have. We will take the Representation with the maximum bitrate if none
+ * is found.
+ * @param {Number} maxBitrate - The maximum bitrate the choosen Representation
+ * should have. We will take the Representation with the minimum bitrate if none
+ * is found.
+ * @returns {Representation|undefined}
+ */
+export default function selectOptimalRepresentation(
+  representations : Representation[],
+  optimalBitrate : number,
+  minBitrate : number,
+  maxBitrate : number
+) : Representation {
+  const wantedBitrate = optimalBitrate <= minBitrate ? minBitrate :
```

Comment:
I'm wondering how we should handle the case where a user has set the min superior to the max.
What should we expect from the `wantedBitrate` value?

Reply:
For the moment, if :
`minBitrate` = 500
`maxBitrate` = 300
`optimalBitrate` = 400 => The `wantedBitrate` will be 500
`optimalBitrate` = 600 => The `wantedBitrate` will be 300
`optimalBitrate` = 100 => The `wantedBitrate` will be 500

Reply:
Yeah, I anticipated that case by letting that behavior undefined but still respecting one of the two limits.
This is even tested in the unit tests for `selectOptimalRepresentation`.

I hesitated to add a warning but that would also perhaps mean I would have to add an explanation of what we do in that case and I prefer to let that case completely undefined so we can do whatever we want with it.

Reply:
Should'nt it be considered an API usage error and throw to warn users that there's probably an inconsistency somewhere in their code ?

Reply:
Yes, throwing might be the better solution, considering that this kind of configuration should never be something wanted by an application and considering that we already throw when other options are incoherent.

I'm working on it.

Reply:
Done